docs/design/queue.md

```diff
@@ -0,0 +1,111 @@
+type Queue struct {
+    metav1.TypeMeta `json:",inline"`
+
+    metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+    // Specification of the desired behavior of a queue
+    // +optional
+    Spec QueueSpec `json:"spec,omitempty" protobuf:"bytes,2,opt,name=spec"`
+
+    // Current status of Queue
+    // +optional
+    Status QueueStatus `json:"status,omitempty" protobuf:"bytes,3,opt,name=status"`
+}
+
+type QueueSpec struct {
+    // The weight of queue to share the resources with each other.
+    Weight int32 `json:"weight,omitempty" protobuf:"bytes,1,opt,name=weight"`
+}
+
+type QueueStatus struct {
+    Unknown int32 `json:"running,omitempty" protobuf:"bytes,1,opt,name=running"`
+    Running int32 `json:"running,omitempty" protobuf:"bytes,2,opt,name=running"`
+    Pending int32 `json:"pending,omitempty" protobuf:"bytes,3,opt,name=pending"`
+    Completed int32 `json:"completed,omitempty" protobuf:"bytes,4,opt,name=completed"`
+    Failed int32 `json:"failed,omitempty" protobuf:"bytes,5,opt,name=failed"`
+    Aborted int32 `json:"aborted,omitempty" protobuf:"bytes,6,opt,name=aborted"`
+}
+```
+
+### QueueController
+
+The `QueueController` will manage the lifecycle of queue: 
+
+1. Watching `PodGroup`/`Job` for status
+2. If `Queue` was deleted, also delete all reated `PodGroup`/`Job` in the queue
+
+### Admission Controller
+
+The admission controller will check `PodGroup`/`Job` 's queue when creation:
+
+1. if the queue does not exist, the creation will be rejected
+2. if the queue is releasing, the creation will be also rejected
+
+### Feature Interaction
+
+#### Customized Job/PodGroup
+
+If the `PodGroup` is created by customized controller, the `QueueController` will count those `PodGroup` into `Unknown` status; because `PodGroup` focus on scheduling specification which did not include customized job's status. 
+
+#### cli
+
+Command line is also enhanced for operater engineers. Three sub-commands are introduced as follow:
+
+__create__:
+
+`create` command is used to create a queue with weight; for example, the following command will create a queue named `myqueue` with weight 10.
+
+```shell
+$ vkctl queue create --name myqueue --weight 10
+```
+
+__view__:
+
+`view` command is used to show the detail of a queue, e.g. creation time; the following command will show the detail of queue `myqueue`
+
+```shell
+$ vkctl queue view myqueue
+```
+
+__list__:
+
+`list` command is used to show all available queue to current user
+
+```shell
+$ vkctl queue list
+Name      Weight  Total  Pending  Running ...
+myqueue   10      10     5        5
+```
+
+#### Security
+
+TODO: who can view, use, delete/update related queues?
+
+#### Scheduler
+
+* Proportion plugin: 
+
+  Proportion plugin is used to share resource between `Queue`s by weight. The deserved resource of a queue is `(weight/total-weight) * total-resource`. When allocating resources, it will not allocate resource more than its deserved resources. 
```

Comment:
IIUC, it seems `total-weight` will sum up all the weight of the queues, as the queue increases, the ` (weight/total-weight)` will change. How do scheduler adjust the proportion of resources an old queue occupies ? Let's say the queue resources are all used up, will this trigger a preemt?

Reply:
Another case, cluster resources may also change

Reply:
Kube-batch will continue executing those actions every X period.